pkg/apis/core/validation/conditional_validation.go

```diff
@@ -17,24 +17,104 @@ limitations under the License.
 package validation
 
 import (
+	"fmt"
+	"net"
+	"strings"
+
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/features"
+	netutils "k8s.io/utils/net"
 )
 
-// ValidateConditionalService validates conditionally valid fields.
-func ValidateConditionalService(service, oldService *api.Service) field.ErrorList {
+// ValidateConditionalService validates conditionally valid fields. allowedIPFamilies is an ordered
+// list of the valid IP families (IPv4 or IPv6) that are supported. The first family in the slice
+// is the cluster default, although the clusterIP here dictates the family defaulting.
+func ValidateConditionalService(service, oldService *api.Service, allowedIPFamilies []api.IPFamily) field.ErrorList {
 	var errs field.ErrorList
 	// If the SCTPSupport feature is disabled, and the old object isn't using the SCTP feature, prevent the new object from using it
 	if !utilfeature.DefaultFeatureGate.Enabled(features.SCTPSupport) && len(serviceSCTPFields(oldService)) == 0 {
 		for _, f := range serviceSCTPFields(service) {
 			errs = append(errs, field.NotSupported(f, api.ProtocolSCTP, []string{string(api.ProtocolTCP), string(api.ProtocolUDP)}))
 		}
 	}
+
+	errs = append(errs, validateIPFamily(service, oldService, allowedIPFamilies)...)
+
 	return errs
 }
 
+// validateIPFamily checks the IPFamily field.
+func validateIPFamily(service, oldService *api.Service, allowedIPFamilies []api.IPFamily) field.ErrorList {
+	var errs field.ErrorList
+	// When ip families are allowed by config or feature gate, validate the combination of clusterIP
+	// and ipFamily. It is required that ipFamily is set on the service internally (by the strategy in
+	// PrepareForCreate|Update or by test code).
+	if oldService != nil && oldService.Spec.IPFamily != nil && service.Spec.IPFamily != nil && *oldService.Spec.IPFamily == *service.Spec.IPFamily {
+		// specifically allow an invalid value to remain in storage as long as the user isn't changing it
+		return errs
+	}
+
+	// Special case empty list and return an error (can be removed when gates are removed, we can panic on nil allowedIPFamilies)
+	if len(allowedIPFamilies) == 0 {
+		if service.Spec.IPFamily != nil {
+			errs = append(errs, field.Invalid(field.NewPath("spec", "ipFamily"), *service.Spec.IPFamily, "may not be set"))
+		}
+		return errs
+	}
+
+	// If the gate is off, setting IPFamily is not allowed
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		if service.Spec.IPFamily != nil {
+			errs = append(errs, field.Forbidden(field.NewPath("spec", "ipFamily"), "may not be set when the gate is off"))
+		}
+		return errs
+	}
```

Comment:
The only scenario I can see where this would return an error is if the feature was disabled, the existing service had the field, and the new service tried to change the value of the field. I expected an "immutable field" error in that case, not a "may not be set" error.

Reply:
Then I should move the immutable check into this method.

Reply:
Ugh, it's mutable if the type is not service type external name.

Reply:
wait... what's mutable?

Reply:
Reorganizing this